drools-benchmarks-parent/drools-benchmarks-reliability/src/main/java/org/drools/benchmarks/reliability/FireAndAlarmBenchmark.java

```diff
@@ -0,0 +1,134 @@
+package org.drools.benchmarks.reliability;
+
+import org.drools.benchmarks.common.DRLProvider;
+import org.drools.benchmarks.common.providers.RulesWithJoinsProvider;
+import org.drools.benchmarks.common.util.BuildtimeUtil;
+import org.drools.benchmarks.common.util.RuntimeUtil;
+import org.drools.benchmarks.reliability.fireandalarm.Alarm;
+import org.drools.benchmarks.reliability.fireandalarm.Fire;
+import org.drools.benchmarks.reliability.fireandalarm.Room;
+import org.drools.benchmarks.reliability.fireandalarm.Sprinkler;
+import org.drools.kiesession.session.StatefulKnowledgeSessionImpl;
+import org.kie.api.conf.EventProcessingOption;
+import org.kie.api.runtime.conf.PersistedSessionOption;
+import org.kie.api.runtime.rule.FactHandle;
+import org.kie.internal.conf.ParallelExecutionOption;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.drools.benchmarks.reliability.AbstractReliabilityBenchmark.Mode.NONE;
+
+@Warmup(iterations = 2000)
+@Measurement(iterations = 1000)
+public class FireAndAlarmBenchmark extends AbstractReliabilityBenchmark{
+
+    private static final String FIRE_AND_ALARM =
+            "import " + Alarm.class.getCanonicalName() + ";" +
+                    "import " + Fire.class.getCanonicalName() + ";" +
+                    "import " + Sprinkler.class.getCanonicalName() + ";" +
+                    "import " + Room.class.getCanonicalName() + ";" +
+                    "rule 'When there is a fire turn on the sprinkler' when\n" +
+                    "  Fire($room : room) \n" +
+                    "  $sprinkler: Sprinkler( room == $room, on == false ) \n" +
```

Comment:
I believe that the problem that I was trying to explain this morning is related with this constraint `room == $room`.

If you disable the option to use an object store with references then after the restore of a failed session this constraint cannot be matched anymore (which is exactly the reason why we introduced the option).

Now I see that this benchmark (and I guess all others) never do a failover, so in this sense the benchmark is correct and ok to be merged since it correctly measures the overhead introduced by the object store with references. That said I wonder if we could add some benchmark that also attempts a failover in order to check the performances of the recovery mechanism. /cc @tkobayas

Reply:
> I believe that the problem that I was trying to explain this morning is related with this constraint `room == $room`.
> 
> If you disable the option to use an object store with references then after the restore of a failed session this constraint cannot be matched anymore (which is exactly the reason why we introduced the option).

Now I understand :)

> Now I see that this benchmark (and I guess all others) never do a failover, so in this sense the benchmark is correct and ok to be merged since it correctly measures the overhead introduced by the object store with references. That said I wonder if we could add some benchmark that also attempts a failover in order to check the performances of the recovery mechanism. /cc @tkobayas

Yes, I will give it some thought and try to create a benchmark with a failover for this opt-in feature.